### DISCUSSION TITLE
ISSG with getServerSideProps

### DISCUSSION THREAD

**Discussion by balazsorban44** [vercel/next.js#19526] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

Our application serves content to subscribed users. To be able to some parts of a page, the user has to be logged in, otherwise they will only be shown a paywall.

## Describe the solution you'd like
I would like to take advantage of Incremental Static Site Generation for pages that require authentication from the user.

To be able to serve content behind a paywall, there are currently two solutions in Next.js

1. Render only a skeleton for a page with `getStaticProps`, and fetch the rest of the page on client-side, after the user has been logged in. There are many good libraries helping with authentication, like [`next-auth`](https://next-auth.js.org). This method sounds good for pages that are highly dynamic, because of requiring a lot of user interaction. (For example dashboards, forms, etc.) But let's suppose some less dynamic pages, that would only need to be changed once in a while (for example when a CMS changes the content of a page).

2. The second method is to fetch the data that needs authentication in `getServerSideProps`. A kind of semi-solution is to create an in-memory cache of these pages, so Next.js doesn't have to go and talk to an API for every request. But it would still require re-rendering pages for every user, even though the content will be the same. One can go further and use something like [examples/ssr-caching](https://github.com/vercel/next.js/tree/canary/examples/ssr-caching), that - if I understand correctly -will cache the entire rendered result, but the user still has to go back to the origin server for this information, and we have to use a custom server.

If the content is public, it can be fetched via `getStaticProps`, and cached in a CDN. Combined with `revalidate` using the stale-while-revalidate strategy, one can create a fast user experience, no matter how many users will try to access the same page.

If the page needs authentication though, there must exist some mechanism that can first check if the user is authorized to look at the page.

I propose such a mechanism, that could make it easy to combine ISSG together with CDNs supporting token authentication, like:
- https://docs.microsoft.com/en-us/azure/cdn/cdn-token-auth
- https://blog.cloudflare.com/token-authentication-for-cached-private-content-and-apis
- https://aws.amazon.com/blogs/networking-and-content-delivery/authorizationedge-using-cookies-protect-your-amazon-cloudfront-content-from-being-downloaded-by-unauthenticated-users/

So for example, when hovering over a link starts to prefetch the page data from the CDN, it would send a cookie that is set, when the user is logged in. If the response is 401, make it easy to the Next.js developer to handle that situation.

## Describe alternatives you've considered

I have considered the above mentioned methods, (1, 2), but none of them seems to be a perfect solution for this use case. We can do better at caching these pages at a CDN and only deliver the data to logged in users, avoiding custom servers, or communicating to the origin server at all. This is just the expansion of all the benefits of ISSG to less dynamic pages that happen to require authentication.

## Additional context

This may be an extension of `getStaticProps`, rather than `getServerSideProps`, I don't have a clear idea/API suggestion yet, and open for comments. What are the strengths and weaknesses of this approach? Can this be achieved already?

**Comment by andrzejwalk** [comment]
I'm also interested in a production-ready solution for this scenario, i.e. how to implement a paywall for static pages.

How to serve:
non-premium static pages -> to non-premium users 
while 
premium static pages -> to premium users.

**Comment by jembach** [comment]
Hi @balazsorban44,

in another discussion I have provided a possible implementation solution: https://github.com/vercel/next.js/discussions/18832#discussioncomment-252786

hopefully someone from the NextJS Team will have a look at this solution to validate if this is possible to implement. I would also help developing the solution.

**Comment by balazsorban44** [comment]
Next.js 12 Middleware is pretty much the implemented version of this. :star_struck:

https://nextjs.org/blog/next-12#introducing-middleware

**Reply by rashidul0405** [reply]
@balazsorban44 Can you please provide more details on the how-to? You can't pass any prop (for example, authenticated) to pages. So how do you then statically maintain/render the paywall and full view of the pages?
